### PR TITLE
WIP: Fix inits doing db IO

### LINKED ISSUE #484: Objects that do a database call in their init
I have run the unit tests of dispersy and profiled them with yappi. I exported the output to callgrind format and exported this into a dotfile using `gprof2dot`.

Using [this code](https://gist.github.com/lfdversluis/6d794e9d27a19a3b75b63ecf1d4b2bb8) I parsed the dotfile to find inits that call the database eventually.

I noticed that a lot of database functions are not covered by the tests such as `executemany`. So this list is most definitely incomplete.

Below the results from this run:
`dispersy/database.py:database:250:DispersyDatabase.execute`
`dispersy/crypto.py:crypto:253:M2CryptoPK.__init__`
`dispersy/util.py:util:146:wrapper`

edit: I think the code can be improved. Doing that now

### PR DESCRIPTION
**Pull request by lfdversluis**
__init__s no longer do DB IO (enhancement)

Fixes #484

### CODE REVIEW

**Comment by lfdversluis** [comment]
@devos50 can you review? (ignore the first commit, it will be dropped).

File: tests/debugcommunity/node.py [hunk 1]
```diff
@@ -30,24 +30,28 @@ class DebugNode(object):
...
         self._dispersy = dispersy
-        self._my_member = self._dispersy.get_new_member(curve)
-        self._my_pub_member = Member(self._dispersy, self._my_member._ec.pub(), self._my_member.database_id)
 
+        self._central_node = None
+        self._tunnel = False
+        self._connection_type = u"unknown"
+
+    @inlineCallbacks
+    def initlialize(self, communityclass=DebugCommunity, c_master_member=None, curve=u"low"):
```
Comment (devos50) [inline tests/debugcommunity/node.py]: There's a typo in the name of the method (should be `initialize`).

File: tests/dispersytestclass.py [hunk 2]
```diff
@@ -109,5 +109,8 @@ def _create_nodes(amount, store_identity, tunnel, communityclass, autoload_disco
         return blockingCallFromThread(reactor, _create_nodes, amount, store_identity, tunnel, community_class,
                                       autoload_discovery, memory_database)
 
+    @inlineCallbacks
     def _create_node(self, dispersy, community_class, c_master_member):
-        return DebugNode(self, dispersy, community_class, c_master_member)
\ No newline at end of file
+        d = DebugNode(self, dispersy)
+        yield d.initlialize(community_class, c_master_member)
+        returnValue(d)
```
Comment (devos50) [inline tests/dispersytestclass.py]: Please add a newline here

File: tests/dispersytestclass.py [hunk 3]
```diff
@@ -109,5 +109,8 @@ def _create_nodes(amount, store_identity, tunnel, communityclass, autoload_disco
         return blockingCallFromThread(reactor, _create_nodes, amount, store_identity, tunnel, community_class,
                                       autoload_discovery, memory_database)
 
+    @inlineCallbacks
     def _create_node(self, dispersy, community_class, c_master_member):
-        return DebugNode(self, dispersy, community_class, c_master_member)
\ No newline at end of file
+        d = DebugNode(self, dispersy)
```
Comment (devos50) [inline tests/dispersytestclass.py]: Minor detail: I would give this variable a more meaningful name.

**Comment by devos50** [comment]
@lfdversluis I gave some comments. Were this all occurrences of inits doing DB I/O btw?

**Comment by lfdversluis** [comment]
What my function found out yes. It turned out that message -> Implementation class was called all the time so many stack traces were gone by fixing that one. This one will be a headache to refactor though once it gets a Deferred, but that's for another PR.

**Comment by lfdversluis** [comment]
@whirm Ready for your review.

File: message.py [hunk 4]
```diff
@@ -522,8 +530,15 @@ def impl(self, authentication=(), resolution=(), distribution=(), destination=()
             distribution_impl = self._distribution.Implementation(self._distribution, *distribution)
             destination_impl = self._destination.Implementation(self._destination, *destination)
             payload_impl = self._payload.Implementation(self._payload, *payload)
-            return self.Implementation(self, authentication_impl, resolution_impl, distribution_impl, destination_impl, payload_impl, *args, **kargs)
+            impl = self.Implementation(self, authentication_impl, resolution_impl, distribution_impl, destination_impl, payload_impl, *args, **kargs)
+            packet = ""
+            if "packet" in kargs:
+                packet = kargs["packet"]
```
Comment (whirm) [inline message.py]: you can do a kargs.get("packet", "")

Comment (lfdversluis) [inline message.py]: Ahh that will be much cleaner 👍

**Comment by whirm** [comment]
Yeah, it's not pretty but I guess it will have to do pending a major refactor...

**Comment by whirm** [comment]
have you run the Tribler tests with this branch?

**Comment by lfdversluis** [comment]
@whirm No, I will do that after I fixed that line you commented on 👍

**Comment by lfdversluis** [comment]
Tribler seems to run happy with this code, although one of the tests errored with a known error.

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
Squash it!

**Comment by lfdversluis** [comment]
@whirm Since the container appears to be unstable, can I somehow run these changes + tribler on BBQ? I would feel more comfortable knowing that the tests pass then. Before a bug was introduced after all and the container hid it.

**Comment by whirm** [comment]
@lfdversluis which container? yours or the test runner?

You can always run your own jobs on jenkins. Just set up one on /pers/

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
@lfdversluis can you rebase this?

**Pull request closed without merging**